Merge and tweak ContributorInfo.rst and GitCheatsheet.rst

lydia-duncan opened the pull request:
Changes based on Michael's proposals via email, as motivated by needing
a place to document the policy decision to add the `[PR]` prefix to mails
asking for reviews.  Deletes GitCheatsheet.rst, and makes
ContributorInfo.rst follow the outline described below:

>  0. Discuss design changes or big development efforts
     - creating a GitHub issue
     - [Design] email
>  1. Set up a branch for development
     - creating a github account
     - forking
     - creating and switching to a feature branch
>  2. Develop and Test contributions locally
     - updating feature branch with changes from main project
     - reading commit messages from main project
     - creating and running new tests
>  3. Request feedback on your changes
     - push your work to your feature branch on github
     - ask for feedback with branch (vs PR) if early
     - create one or more pull requests
     - [PR] email
     - work with your reviewers
>  4. Pass relevant testing
     - "full local testing"
>  5. Merge changes in to master
     - checklist before merging:
        - design changes discussed?
        - PR reviewed?
        - contributor agreement signed?
        - history in PR does not contain anything not to be in repo?
        - relevant configurations pass testing?
     - reviewer merges generally, for external contributors
>  6. Watch automatic testing & address issues
     - smoke test
     - nightly testing

>  HOWTO/Git/GitHub Details
>   - Stuff linked from above
>   - Other Git Commands

>  Policy Details
>   - Stuff linked from above
>   - Information for Reviewers

lydia-duncan commented:
@mppf - this is finally ready for review, sorry it took so long.  I recommend clicking the "view" button to take a look at the file as it stands now rather than looking at the raw contents

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,80 +9,297 @@ The git repository for the project is hosted on Github at
+  Next, put in a message to your reviewer about the purpose of your pull request
+  and give the pull request a useful title. It's a good time to draft the commit
+  message that you will need when merging the pull request.  See `Final merge
+  message`_ for recommendations on what that commit message should look like.
```
Can we add "It is a good practice to make the PR description match the final merge message. Some developers copy the PR description into the merge message and then wrap any long lines."

mppf replied:
Or maybe just say something here like "Your PR message will introduce the changes to reviewers and form the basis for the merge message".

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -205,9 +423,333 @@ Mail chapel-developers@ for review and merge
+.. _Final merge message:
+
+Final merge message
++++++++++++++++++++
```
I think I'd like this section to treat both the Final PR message and the Final merge commit message.

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,80 +9,297 @@ The git repository for the project is hosted on Github at
+   #. `Find a reviewer`_
+   #. `Work with your reviewers`_
+
 #. `Test changes`_ (can be done by team at Cray)
```
I think we should remove "can be done by team at Cray". Also, the original outline (which is in your PR message) doesn't have this as a top-level item, and instead lists it under a checklist of things to do before merging. I think I'd still be in favor of moving it under Merge changes to master or renaming the section "Getting Ready to Merge" and include it as part of a checklist. The full checklist is in your PR message. I think I'd prefer the checklist exist in outline form at the start of "Getting Ready to Merge". Most of it is covered currently in the long-sentences form, but a checklist can also be a quick guide.

lydia-duncan replied:
Actually, the checklist does have it as a top-level item before merging - it's bullet 4.  That's why I put it in both spots.  I can change that, though, if you think this is a better strategy

lydia-duncan replied:
(deleted my comment because it was wrong)

lydia-duncan replied:
deleted my comment on deleting the comment.  I was right the first time.  Testing is mentioned twice in the outline above, both times at the top level (bullet 2 and bullet 4).  I think it is fair to move the second mention under the Getting Ready to Merge section, though

lydia-duncan replied:
Three times.  I'm tired today.

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,80 +9,297 @@ The git repository for the project is hosted on Github at
-These are expected to evolve over time as the governance of Chapel is migrated
-from Cray to an external/community body (the major elements are likely to be
-similar, though the specific people involved are likely to change and grow).
+Before performing major work on the language, please open an issue on the git
```
"performing major work on the language" -> "embarking on a big project"

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,80 +9,297 @@ The git repository for the project is hosted on Github at
-* If you will need commit/push access to the main repository,
-  `chapel-lang/chapel`_, send a request including your github username to
-  chapel_admin _at_ cray.com.
+This consists of two sets of operations.  The first (`Set up repository`_) only
```
I think I'd call it "Fork the main repository" since "Fork" is the term the GitHub documentation uses. https://help.github.com/articles/fork-a-repo/

lydia-duncan replied:
It originally was named that, but I changed it to distinguish the description of the set of actions needed only once (creating an account, getting your environment correct, etc.) from the actual operation of hitting the fork button (which had a very similar name).

mppf replied:
Since creating an account etc isn't technically "setting up a repository" either, maybe you should name the section something like "first steps" or "getting set up" or something more generic, and then consider "fork a repo" as a task within that?

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,80 +9,297 @@ The git repository for the project is hosted on Github at
-    You do not need commit/push access to the main repo in order to contribute
-    code. Reviewers on the core team can pull, review, and merge your pull
-    requests.
+Set up repository
```
Let's generally move the "Set up a repository" and "Create a new branch" sections down to the HOWTO/Git/GitHub details section.

lydia-duncan replied:
Okay, I'm having a bit of trouble with this one, so I figured I should ask.  I see what you mean about the first three bullets, but they are already mostly just links and I think the remainder should be at that location still.  I moved a couple of the bullets around, since "who has commit access" and "third party code" weren't actually steps.  After the modifications that have happened, what would you prefer I do?

mppf replied:
It's going better. I think the next thing I think we should do with this section is to consider making "Getting set up" and "Create new branch" just different top-level sections. Mostly because the 1st one only happens once.

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,80 +9,297 @@ The git repository for the project is hosted on Github at
+
+   Note: You do not need commit/push access to the main repo in order to
+   contribute code. Reviewers on the core team can pull, review, and merge your
+   pull requests.
```
Can we move this part down into a Policy Details section, e.g. "Who has commit access to the main repository?".
In the "Set up repository" (or fork) section, it suffices to say that you don't need commit access in order to proceed.

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,80 +9,297 @@ The git repository for the project is hosted on Github at
+   work is not being pursued redundantly, etc.  Similarly, fundamental changes
+   to the language or architecture should be circulated to the
+   chapel-developers_ and/or chapel-users_ lists to make sure effort is not
+   wasted.
```
This is already covered under "Discuss design changes or big development efforts". I'd just remove this section. You might consider adding to "Discuss design changes or big development efforts" that it is sometimes appropriate to raise user-facing changes on chapel-users. (and if you want to describe when that is appropriate, it'd be in a a section at the bottom, if it's more than a sentence or two).

lydia-duncan replied:
I disagree about removing this section - I've found that repeating concepts in a document when relevant helps readers to remember them, and this is something we want them to remember.

I will add the user-facing information you describe, though.

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,80 +9,297 @@ The git repository for the project is hosted on Github at
+#. If your work will require committing any third-party code that you are not
+   developing yourself (or code that you've developed as a standalone package),
+   alert the chapel-developers_ mailing list of this as, presently, such code
+   packages must be approved by Cray leadership before being committed.
```
The part about third-party code should just be mentioned here ("Third-party code requires additional approvals") and link to a section in Policy details at the bottom.

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,80 +9,297 @@ The git repository for the project is hosted on Github at
+#. Sign a Chapel contributor's agreement and mail it, with your GitHub
+   ID, using the instructions here:
+
+   https://github.com/chapel-lang/chapel/tree/master/doc/rst/developer/contributorAgreements/
```
Is this the right place to bring up the contributor agreement? We could also request it at the time the PR is created, which is where we check in practice. Here, it might suffice to say something like "You will need to sign a contributor agreement for your changes to be accepted" and then refer to the link.

lydia-duncan replied:
I think mentioning it before the work is started is a good idea, especially because institution contributor agreements can take a long time to get filled out - I think it is better for the contributor to know they need to start this process so it can be in motion while they work, rather than have them discover the details when they are asking for review and have the PR languish because of it (as has happened in the past)

lydia-duncan replied:
If it isn't explicitly mentioned at the PR creation section, I'll make sure it gets mentioned, though.

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,80 +9,297 @@ The git repository for the project is hosted on Github at
+
+    git checkout -b <branch_name>
+
+  Using a concisely named branch is encouraged.
```
I don't think we care about branch names. (If we keep that text, it needs to drop leading spaces).

Also, I'd consider putting the actual command name in the HOWTO section.

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,80 +9,297 @@ The git repository for the project is hosted on Github at
+merge message once the pull request is going in. Likewise, it is OK to have many
+small commits that reflect the history of development rather than commits for
+the feature.  See `Development commands`_ for how to perform some common
+operations during development.
```
This paragraph is good and basically works how I was hoping (top-level description of outlined topic, refers later for details)

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,80 +9,297 @@ The git repository for the project is hosted on Github at
+operations during development.
+
+As you work, you will want to periodically bring in changes from the main Chapel
+project to your feature branch:
```
I don't understand why this section and the commands below are here and in a linked-to section in HOWTO.
If you think it's important enough to call out this operation (vs all of the other things a developer might do with git), let's just make a sentence that refers to a specific section within Development Commands for this operation.

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,80 +9,297 @@ The git repository for the project is hosted on Github at
+commit`` to finish the merge process.
+
+If you want to understand the changes that occurred upstream, see `Read commit
+messages`_ below.
```
This is a broken link

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,80 +9,297 @@ The git repository for the project is hosted on Github at
+
+.. _Test your feature:
+
+Test your feature
```
Maybe this should just be "Adding new tests"?

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,80 +9,297 @@ The git repository for the project is hosted on Github at
+
+You will probably need to create new tests for your feature. See `Test System`_
+for more information on this process.  To do the most basic testing, you would
+do one of:
```
I don't think we should include these details here. I'm talking about everything after "To do the most basic testing, you would". Isn't the link to Test System sufficient?

Maybe we could keep just the "Any addition/change to the Chapel test system should pass testing.." paragraph?

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,80 +9,297 @@ The git repository for the project is hosted on Github at
+Contributors should be reasonably confident in the testing done on their code
+before asking for a review.  Should additional testing resources be needed, you
+can request help from a member of the core Chapel team when creating your pull
+request.
```
This should be in the section "Submit pull request" if anywhere?

I'm not sure I agree with the sentiment though... but I might be in the minority.

I think that WIP pull requests are reasonable and not a significant burden to the project. In fact a WIP pull request is nicer than just getting comments on a branch because:
 * it's easy to link to (e.g. from issues, other PRs)
 * it leaves a history of any feedback comments somewhere people might find them.

The only thing about WIP pull requests is that they should generally be closed until the PR is actually ready for review, final testing, and merging. That way, our PRs list does not get too big.

lydia-duncan replied:
I think WIP pull requests are fair (and maybe should be called out as explicitly okay).  These lines are more meant to dissuade interactions like "hey, review my new test?" when start_test hasn't been run over the new test prior to asking and doing so would find an obvious issue (forgot a .good file, fails compilation when the test is expected to execute, there's an extra newline in the expected output, etc).

lydia-duncan replied:
Maybe I should be more specific, and divide it into categories?  For instance, if you think the code is ready to go, it should be tested before asking for review; if you are looking for advice on a implementation direction, having tested it previously isn't as necessary, etc?

lydia-duncan replied:
Being sure to distinguish between "asking for review" and "opening a pull request", of course

mppf replied:
Personally, I'd just focus the efforrt in this document on requesting that the reader clearly communicate to their reviewer the state of it - and point out that if it's doesn't work at least a little bit, getting feedback might not be that helpful. E.g. suggest that the PR message indicate major TODOs (like "check new tests work").
I'm happy with the idea of encouraging people to test to their satisfaction before making a PR, but I think it's a complicated issue enough that it might be misintepreted to exclude the way I use PRs (for example).

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,80 +9,297 @@ The git repository for the project is hosted on Github at
+
+.. _Testing guidance:
+
+Testing your patch:
```
This section (everything after "Testing your patch") belongs in the Policy Details section. I'm not sure if it makes sense to link to it from here at all though. I'm leaning towards only linking to it from the "Getting ready to merge a PR" section that I'm asking for.

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,80 +9,297 @@ The git repository for the project is hosted on Github at
+feature branch, pushing your work to that feature branch will
+update the pull request.
+
+This will enable others to see your work.
```
I think this whole section should be shorter & refer to the HOWTO for the details. How about

"Pushing your changes to your feature branch on GitHub will enable others to see your work.
Note that if you have already created a pull request from a feature branch, pushing to that feature branch will update the pull request."

and then link below for details on how to actually push.

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,80 +9,297 @@ The git repository for the project is hosted on Github at
+++++++++++++++++++++++++++++++++++++++++++++++++
+
+Not ready to merge yours changes, but still want to see if your work is going
+in the right direction?  Feel free to ask for early feedback!
```
That's fine, but it doesn't say *how* to ask for early feedback which is almost as important in this document as saying it's a good idea. Here are some possibilities for sharing the work:
 * point somebody to your feature branch on github
 * create a WIP PR for your feature branch and close it
Here are some ways to discuss:
 * discuss in the WIP PR
 * discuss in a GitHub issue
 * discuss on chapel-developers
 * discuss privately

Which of these do we recommend? (I personally recommend the closed WIP PR strategy, but I don't know if others would be uncomfortable with that).

lydia-duncan replied:
Personally, I'm not as concerned with the how, but it seems reasonable to bring up the various strategies.

It might discourage contributions/interactions to require a WIP PR instead of allowing that option and pointing someone to just the branch - some people might be okay showing one person their unfinished branch but not having it visible on main repo until it is in a better state, for instance.

lydia-duncan replied:
Is this something that we should get broader input on?

mppf replied:
I absolutely don't think we should require a WIP PR for this sort of thing. I think it would make sense for the document to describe though, since it might not be obvious what etiquette is important around WIP PRs.

I think we should probably mention WIP PRs as their own topic and also point out that you can get feedback from anybody who can see your code however you like. Maybe something like "If you already know how to get feedback, that is fine. If not, you could create a WIP PR or email chapel-developers". E.g.

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,80 +9,297 @@ The git repository for the project is hosted on Github at
-* Develop your feature, bug fix, etc on your fork. Using a concisely named
-  branch is encouraged.
 
 .. _Submit pull request:
```
The way the above section flows, it seems like submitting a PR is a good way to ask for early feedback. If that's not the case, I think it needs some adjustment somehow.

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,80 +9,297 @@ The git repository for the project is hosted on Github at
-* In order for a pull request to be accepted and merged, it should a) pass
-  testing and b) be reviewed by a member of the core Chapel team (currently, a
-  member of the Cray Chapel team).
+  To do this, after pushing your changes to your feature branch on GitHub,
```
All these details belong in the HOWTO section and should be linked to from here.

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,80 +9,297 @@ The git repository for the project is hosted on Github at
+  message that you will need when merging the pull request.  See `Final merge
+  message`_ for recommendations on what that commit message should look like.
+
+  You will have to have signed a contributors agreement.
```
If this links to the contributor agreement readme, I think this detail could be removed from earlier.

lydia-duncan replied:
Eh, if I were reading the documenting and got reminded here that I needed to fill out the agreement, I'd rather not have to scroll/search the document to find the link I need

mppf replied:
Then put the link in both places. I'm mostly trying to get this top section to be shorter / faster read.

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,80 +9,297 @@ The git repository for the project is hosted on Github at
+
+  a) pass testing and
+  b) be reviewed by a member of the core Chapel team (currently, a member of the
+     Cray Chapel team).
```
I'd lean towards putting this in to "Getting ready to merge a PR" section

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,80 +9,297 @@ The git repository for the project is hosted on Github at
+  b) be reviewed by a member of the core Chapel team (currently, a member of the
+     Cray Chapel team).
 
 * It is considered good practice to keep PRs (pull requests) to a reasonable
```
I think this section should move down to Policy and just be linked to there with a 1-liner like "It's a good idea to keep PRs reasonably sized".

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -102,91 +319,92 @@ Submit pull request
+  prior to the subject of your email.  Even the developers that have write
+  access to the Chapel repository need to have all non-trivial changes
+  reviewed. Developers who have been given write access can merge trivial
+  changes (e.g. small bug fixes, documentation changes) without review.
```
Using the @ mention syntax in a comment on the PR is a best practice to request review from a particular person. People without commit access can't use the "Request reviewer" pane on the right. I think @ mentioning is better than emailing separately if you know who the reviewer is.

I'd like this section to be shorter. E.g.

"Once your PR is ready, you'll need to request a review. If you know who you'd like to review it, @ mention them in a comment on the PR and ask them to have a look. If you don't know who should review it, send an email to the chapel-developers_ list requesting a review and linking to the PR. Such an email should have a subject line starting with `[PR]`."

The bit starting with "Even the developers..." belongs in a "Who can commit" section in Policy Details. I don't think it's necessary here at all. The Policy Details section can say that people who have commit access can commit trivial changes without review & describe what counts as a trivial change.

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -102,91 +319,92 @@ Submit pull request
+  chapel-developers_) into the conversation to get a broader opinion (one of
+  the jobs of the reviewer is to serve as a proxy for other developers, or to
+  bring those developers into the conversation if they feel unqualified to do
+  so).
```
This seems pretty good. However I think it could point out two more things:
 * one should generally try to do whatever their reviewer asks
 * sometimes a reviewer will ask for something really hard. Just try to make sure they understand the magnitude of the request & try to discuss if it's really necessary to do before merging.

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -102,91 +319,92 @@ Submit pull request
+If you did not have the resources to perform at least a full correctness test
+run, this is the point at which a reviewer would do so for you.  See
+`Testing guidance`_ above for more details on what an appropriate amount of
+testing is.
```
I think there are three things to link to from this section:
 1. The earlier section, basically saying "You already ran some testing, right"?
 2. The Policy section, saying PRs should pass at least full local testing, talking about configurations etc etc
 3. The HOWTO section or the Test System document, to describe the mechanics of running tests

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -102,91 +319,92 @@ Submit pull request
   between the two of them.
 
-  .. note::
+  Note: Recall that while git history can be rewritten, it is difficult and does
```
I'd probably just say "It's not generally possible to completely remove a commit from git by the time it makes it in to the master branch. So be very careful...".

I think this should be in the "Getting Ready to Merge" section

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -102,91 +319,92 @@ Submit pull request
+  privileges, please look through the `Reviewer responsibilities`_ once more
+  before merging the change.
+
+After the final version of the change has been agreed upon, navigate to the
```
This part on the mechanics of merging a PR belongs in the HOWTO section.

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -102,91 +319,92 @@ Submit pull request
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
 * In short order, a smoke-test will be run against the commit to make sure that
   nothing basic has been broken by it.  Monitor the
```
The part about "or that they are due to a commit other than yours" could be simplified to "to ensure that your commit has not caused failures"

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -205,9 +423,333 @@ Mail chapel-developers@ for review and merge
   have intentionally made the process as lightweight as possible, so it can be
   used for changes big and small alike.
```
This last bullet no longer belongs in this section. Not sure it's really necessary anywhere.

mppf commented:
@lydia-duncan - thanks for putting this together. I'm done offering feedback for the moment. I'll look at the HOWTO and Policy sections in more detail for the next revision, but I'd like to get the main section shorter & better first. (So I'm saying I'd like to look at both parts again after some changes).

lydia-duncan commented:
Thanks for the review!  I'll reping you once I've fully incorporated your suggestions

lydia-duncan commented:
I hate verifying links.  I'm going to wait and fix them all at once right before merging instead of after every change because it is just getting annoying.

lydia-duncan commented:
I left the branch name thing alone, mostly because I haven't decided if I care or not (like, it's good to have short names for the folks without tab completion that want to test the change, but it isn't a big deal).  Other than that and the question about how to rearrange "Getting set up", I think I have addressed all your feedback and this is ready for a next review - let me know if you noticed I missed something.

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,184 +9,274 @@ The git repository for the project is hosted on Github at
+subject header for discussion with the community and to ensure that you are
+aware of any parallel efforts in that area.  If the change will impact the user
+experience, it would be appropriate to additionally send this email to
+chapel-users_.
```
I think I'd word this less strongly - something like "It may also be good idea to ask the user community (on chapel-users_, say) what they think about a project or language direction".
I don't think we want to suggest that *all* language changes (even) should have emails to chapel-users.

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,184 +9,274 @@ The git repository for the project is hosted on Github at
+   https://github.com/chapel-lang/chapel/tree/master/doc/rst/developer/contributorAgreements/
+
+* You do not need commit/push access to the main repo in order to
+  contribute code.  See `Who has commit access`_ for questions.
```
Let's just say "See `Who has commit access`_."

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,184 +9,274 @@ The git repository for the project is hosted on Github at
-    creativity, and may not be feasible at all, due to the intertwined nature
-    of the code.
+You will probably need to create new tests for your feature. See `Test System`_
+for more information on this process.
```
Can we also link to "Creating a Simple Test" in Test System? That's the information a beginner might immediately need.

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,184 +9,274 @@ The git repository for the project is hosted on Github at
-  * Any addition/change to the Chapel test system should pass testing when that
-    test/directory is run with ``start_test`` (and performance tests should
-    also pass testing for ``start_test -performance``).
+Push your changes to your feature branch on GitHub will enable others to see
```
"will enable" -> "to enable"
(or other fix for typo)

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,184 +9,274 @@ The git repository for the project is hosted on Github at
+- a separate Github issue
+- the chapel-developers_ mailing list
+- a private communication
+- some other strategy agreed upon by all involved parties
```
The `-` formatting isn't making a list here. But it might be better as some sentences anyway (takes less space).

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,184 +9,274 @@ The git repository for the project is hosted on Github at
+  (individually or via chapel-developers_) into the conversation to get a
+  broader opinion (one of the jobs of the reviewer is to serve as a proxy for
+  other developers, or to bring those developers into the conversation if they
+  feel unqualified to do so).
```
The parenthetical here seems too long - maybe it should be its own sentence?

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,184 +9,274 @@ The git repository for the project is hosted on Github at
+  feel unqualified to do so).
+
+
+.. _Test changes:
```
I (still?) think Test Changes should be under Getting Ready to Merge

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,184 +9,274 @@ The git repository for the project is hosted on Github at
+- the PR has been reviewed
+- the contributor agreement has been signed
+- the history in the PR does not contain anything not to be in the repo
+- relevant configurations pass testing
```
I like this one as a bulleted list, but the formatting isn't working now. Also, could it use the word "checklist" in the  topic sentence?

lydia-duncan replied:
I keep forgetting that you need a buffer line for lists to format correctly

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,184 +9,274 @@ The git repository for the project is hosted on Github at
+- all design changes have been discussed
+- the PR has been reviewed
+- the contributor agreement has been signed
+- the history in the PR does not contain anything not to be in the repo
```
I thought this would link to a policy section something below? We had about a paragraph about this earlier.

mppf replied:
Oh, I see that paragraph is in the "it can be merged' section below. I think it should move down.

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,184 +9,274 @@ The git repository for the project is hosted on Github at
+  the time it makes it in to the master branch. So be very careful not to commit
+  anything that you might regret later (e.g., sensitive code, code owned by
+  anyone other than yourself). Ideally, the review will catch such issues, but
+  the ultimate responsibility is on the developer.
```
I think a good sign this belongs in the Policy Details section with a link from this section is that it starts with "Note:". Let's move it down and link to it.

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -9,184 +9,274 @@ The git repository for the project is hosted on Github at
+
+* If you are reviewing code from an external contributor without push
+  privileges, please look through the `Reviewer responsibilities`_ once more
+  before merging the change.
```
My hope/intent is that the checklist above covers this part and would be easier for one of us to actually look through. If you agree, maybe refer to the checklist above?

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -195,19 +285,548 @@ Mail chapel-developers@ for review and merge
+Configure your local git
+++++++++++++++++++++++++
+
+Here is the uncommented version of the commands:
```
I don't understand what "the uncommented version of the commands" is supposed to mean. There are comments in there.

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -195,19 +285,548 @@ Mail chapel-developers@ for review and merge
+    git add path/to/file
+
+    # (sort of) similar to:
+    svn add path/to/file
```
I think it'd be an improvement to put the svn commands in the comment, like
`# (sort of) similar to: svn add path/to/file`
There are a lot more of these if you agree.
My thought here is that one might expect to be able to paste whatever shows up in the code box, and the svn command doesn't help in that context.

lydia-duncan replied:
That seems reasonable!

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -195,19 +285,548 @@ Mail chapel-developers@ for review and merge
+
+If there are conflicts, you will be asked to resolve them. Once the affected
+files have been fixed, stage them with ``git add``, and then call ``git
+commit`` to finish the merge process.
```
You don't have to do this now, but I think it'd be nice to cover rebasing with master and also rebasing to change history in this document.

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -195,19 +285,548 @@ Mail chapel-developers@ for review and merge
+    git commit --amend
+
+Note: This should only ever be done to commits that **have not been pushed** to
+a remote repository.
```
I actually disagree about this. It's OK to do with commits that have only been pushed to your feature branch. I do it regularly that way. You just have to `git push -f` when you send them up to GitHub.

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -195,19 +285,548 @@ Mail chapel-developers@ for review and merge
+    git reset --soft HEAD~1
+
+Note: This should only ever be done to commits that **have not been pushed** to
+a remote repository.
```
Maybe all of these "modifying history" operations could have their own subsection here? And then we could start with some caveats of doing so? And maybe rebasing with waster could be in it too?

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -195,19 +285,548 @@ Mail chapel-developers@ for review and merge
+    # if you forgot your branch name, you can get it by running
+    git branch
+
+    # it is the starred one...
```
This could say that `push -f` is necessary if you've modified history on your feature branch (but see modifying history section)

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -195,19 +285,548 @@ Mail chapel-developers@ for review and merge
+merge the pull request from the command line also and the pull request page has
+details). When you click "Merge pull request", you will need to enter a commit
+message. See `Final merge message`_ for a reminder on what that commit message
+should entail.
```
The link here is good, but the text doesn't really say that we're generally copying the PR message in to the merge message box.

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -195,19 +285,548 @@ Mail chapel-developers@ for review and merge
+
+(If you use this command on a feature branch, you'll just be updating to the
+latest work stored on github. See `Develop and test contributions locally`_ for
+how to update a feature branch with new changes from the main Chapel project)
```
I think this part might already be covered in "Bring in changes from the main Chapel project"

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -195,19 +285,548 @@ Mail chapel-developers@ for review and merge
+
+It is perfectly acceptable to abandon such PRs (especially in favor of a cleaned
+up version of the code) when the git history becomes too large, so long as a
+link to the original PR is provided when the change is eventually merged.
```
Looks good, thanks. I might say something about how linking to the original PR is good, to preserve the discussion. I'm not sure it's a "rule" at the same level of having signed a contributor agreement...

lydia-duncan replied:
I thought I had included that, but maybe it isn't as clear as it should be.  I'll re-examine it

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -195,19 +285,548 @@ Mail chapel-developers@ for review and merge
+It is good practice for the pull request title to match the topic line, and the
+detailed explanation to match the pull request description.  For contributors
+without merge permissions, this will allow the reviewer to merge changes with
+minimal additional support.
```
This seems a little odd, structurally. Contributors without merge permissions won't be writing a merge message, so why would they be reading this section? Maybe we just need to be clearer that the reviewer will do the merge message. Not sure how to fix it best.

mppf commented:
@lydia-duncan looks good, I had some feedback but I'm comfortable with this being merged any time. I don't think we added any guidance that would be *surprising* to other team members but if you know of anything along those lines we should bring them up. Thanks!

mppf commented on doc/rst/developer/bestPractices/ContributorInfo.rst:
```diff
@@ -452,23 +430,32 @@ commit`` to finish the merge process.
++++++++++++++++++++++++++
+
+The following commands are **unsafe** if the changes they overwrite have been
+pulled by other repositories.
```
I'm not sure I'd use "unsafe" here - I'd just say something like "Changing git history can cause problems if the commits have been pulled to other repositories" or something along those lines.
